server/src/main/java/org/opensearch/wlm/tracker/QueryGroupResourceUsage.java

```diff
@@ -0,0 +1,128 @@
+     */
+    public double getCurrentUsage() {
+        return currentUsage;
+    }
+
+    public void setCurrentUsage(double currentUsage) {
+        this.currentUsage = currentUsage;
+    }
+
+    public static QueryGroupResourceUsage from(ResourceType resourceType) {
+        if (resourceType == ResourceType.CPU) {
+            return new QueryGroupCpuUsage();
+        } else if (resourceType == ResourceType.MEMORY) {
+            return new QueryGroupMemoryUsage();
+        }
+        throw new IllegalArgumentException("Invalid resource type: " + resourceType + ". It is currently not supported in wlm");
+    }
+
+    /**
+     * Determines whether {@link QueryGroup} is breaching its threshold for the resource
+     * @param queryGroup
+     * @return whether the query group is breaching threshold for this resource
+     */
+    public boolean isBreachingThresholdFor(QueryGroup queryGroup, WorkloadManagementSettings settings) {
+        return getCurrentUsage() > getNormalisedThresholdFor(queryGroup, settings);
+    }
+
+    /**
+     * returns the value by which the resource usage is beyond the configured limit for the query group
+     * @param queryGroup instance
+     * @param settings {@link WorkloadManagementSettings} instance
+     * @return the overshooting limit for the resource
+     */
+    public double getReduceByFor(QueryGroup queryGroup, WorkloadManagementSettings settings) {
+        return getCurrentUsage() - getNormalisedThresholdFor(queryGroup, settings);
+    }
+
+    /**
+     * initialises the member variable currentUsage
+     * @param tasks list of tasks in the query group
+     * @param timeSupplier nano time supplier
+     */
+    public void initialise(List<QueryGroupTask> tasks, Supplier<Long> timeSupplier) {
+        this.setCurrentUsage(this.calculateResourceUsage(tasks, timeSupplier));
+    }
+
+    /**
+     * normalises configured value with respect to node level cancellation thresholds
+     * @param queryGroup instance
+     * @param settings {@link WorkloadManagementSettings} instance
+     * @return normalised value with respect to node level cancellation thresholds
+     */
+    public abstract double getNormalisedThresholdFor(QueryGroup queryGroup, WorkloadManagementSettings settings);
+
+    /**
+     * calculates the current resource usage for the query group
+     * @param tasks list of tasks in the query group
+     * @param timeSupplier nano time supplier
+     */
+    protected abstract double calculateResourceUsage(List<QueryGroupTask> tasks, Supplier<Long> timeSupplier);
+
+    /**
+     * class to store cpu usage for the query group
+     */
+    public static class QueryGroupCpuUsage extends QueryGroupResourceUsage {
+        @Override
+        public double getNormalisedThresholdFor(QueryGroup queryGroup, WorkloadManagementSettings settings) {
+            return settings.getNodeLevelCpuCancellationThreshold() * queryGroup.getResourceLimits().get(ResourceType.CPU);
+        }
```

Comment:
Ideally, `WorkloadManagementSettings` should not be method parameter, but something instance method should have access to. Otherwise, this is very much like Helper/Util method

Reply:
I believe keeping this as part of object state is only going incur unnecessary overhead since these values are only needed for the calculation but will remain throughout the run of the logic (QueryGroupService). Because this reasoning will lead to having all method parameters as object state.

But I believe instead of the settings maybe we can just pull the **normalisedThreshold** from the client of this method.

Reply:
I have refactored these into utility classes which does not carry state